File: search_sorted_array_uknown_size.py
## Time complexity O(log N)
## Space complexity O(1)

# """
# This is ArrayReader's API interface.
# You should not implement it, or speculate about its implementation
# """
#class ArrayReader:
#    def get(self, index: int) -> int:

class Solution:
    def search(self, reader: 'ArrayReader', target: int) -> int:
	
        
        left = 0
        right = 1
        
        while reader.get(left) <= target and (not (reader.get(left) <= target and target <= reader.get(right))):
            left = right
            right = right * 2
            
        
        while left <= right:
            mid = (left + right) // 2
            
            x = reader.get(mid)
            if x == target:
                return mid
            elif x < target:
                left = mid + 1
            else:
                right = mid - 1
               
            
        return -1
            

File: Search_Sorted_array_uknown_size.java

// Time Complexity :O(log n ) 
// Space Complexity : O(1)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : N/A


class Solution {
    public int search(ArrayReader reader, int target) {
        int l = 0, r=1;
        int mid;

        //finding max size
       while (reader.get(r) < target){
           //adding l= r missing condition
           l = r;
           r = r * 2;
              }

        //finding minimum
        while(l<=r)
        {
            mid = (l+r)/2;
            if (reader.get(mid) == target) {
                return mid;
            }
            if (reader.get(mid) >= target) 
            r = mid-1;
            else 
            l = mid+1 ;      
        }
         return -1;

}}
